hardware-handler/nightmare_movement_engine/src/modules/robot_math.py

```diff
@@ -1,43 +1,47 @@
 import numpy as np
-from numpy import sin, cos, tan, arccos, arcsin, arctan, arctan, arctan2, sqrt
-from modules.config import legs, DEFAULT_POSE, SERVO_OFFSET, POSE_OFFSET, POSE_REL_CONVERT
+from numpy import sin, cos, arccos, arctan2, sqrt
+from .config import legs, SERVO_OFFSET, POSE_OFFSET, POSE_REL_CONVERT
 
 PI = np.pi
+EPSILON = 0.001
+
+
+def abs_ang2pos(ang):
+    angles = np.reshape((ang - SERVO_OFFSET)[:-1], newshape=(6, 3))
+    rel_pose = np.array([rel_ang2pos(ang, leg.dim) for ang, leg in zip(angles, legs)])
+    pose = (rel_pose * POSE_REL_CONVERT) + POSE_OFFSET
+    return pose
 
 
 def abs_pos2ang(pose):
     rel_poses = (pose - POSE_OFFSET) * POSE_REL_CONVERT
     angles = np.zeros(shape=(6, 3))
-    for i, rel in enumerate(rel_poses):
-        angles[i] = rel_pos2ang(rel, legs[i].dim)
+    angles = np.array([rel_pos2ang(rel, leg.dim) for rel, leg in zip(rel_poses, legs)])
     return np.append(angles.ravel(), 0) + SERVO_OFFSET
 
 
 def rel_pos2ang(rel_pos, leg_dim):
-    x = rel_pos[0]
-    y = rel_pos[1]
-    z = rel_pos[2]
-    CX = leg_dim[0]
-    FM = leg_dim[1]
-    TB = leg_dim[2]
-    d = sqrt(x**2 + y**2)
-    d1 = d - CX
-    d2 = sqrt(z**2 + d1**2)
+    x, y, z = rel_pos
+    if(z < EPSILON and z > -EPSILON):  # check if z is zero because this could cause division by zero or nan!!!
```

Comment:
simplify this to `-EPSILON < z < EPSILON`

Reply:
fixed